### DISCUSSION TITLE
Как создать multiarch тег для Docker образа

### DISCUSSION THREAD

**Discussion by temikfart** [CoolStory-Studio/coolstory-gram#11] [Guides]
Столкнулся с проблемой, что мой макбук использует архитектуру `arm64`, тогда как более популярной является `x86_64`. Поэтому нам придется поддерживать обе архитектуры. В связи с этим возникла проблема создания docker образа, основанного на Linux системе, с установленной библиотекой Boost C++.

Если вы попытаетесь запушить на Docker Hub какой-нибудь тег в ваш репозиторий, то вы столкнетесь с проблемой, что он будет только под вашу архитектуру (если не указывать специальную опцию `--arch` в команде `docker build`). Может показаться, что 1 тег = 1 архитектура и нужно наплодить теги, но я увидел на том же Docker Hub'е, что есть теги с несколькими архитектурами сразу. Поэтому немного погуглив нашел информацию, что каждый docker-образ хранит в специальном **манифесте** информацию о том, на какой архитектуре он должен быть запущен.

Немного погуглив я нашел неплохой гайд по созданию таких **multiarch**-тегов. Вот [она](https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/).

Если коротко, то нужно выполнить следующий алгоритм:
1. Создать контейнер для утилиты `docker buildx` - специальной экспериментальной утилиты Docker'а для сборки таких образов:
```shell
docker buildx create --use
```
2. Написать какой-нибудь Dockerfile основанный на таком же multiarch-теге, например, `ubuntu:22.04`, который умеет одновременно и в `arm64` и `amd64`. Пусть это будет добавление библиотеки Boost:
```Dockerfile
FROM ubuntu:22.04

RUN apt update && \
        apt upgrade -y && \
        apt install libboost-all-dev
```
3. После этого собрать его с утилитой `docker buildx` указав необходимые архитектуры, репозиторий и тег:
```shell
docker buildx build --push --platform linux/amd64,linux/arm64/v8 --tag coolstory/ubuntu-boost:1.82 . 
```
_Примечание: не забудьте удостовериться, что репозиторий `coolstory/ubuntu-boost` создан и у вас есть к нему доступ._

Таким образом, на Docker Hub будет загружен Docker образ с тегов на несколько архитектур сразу.

**Comment by cann3v** [comment]
найс гайд bro
